src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -24,40 +24,27 @@ public override BoundNode VisitDeconstructionAssignmentOperator(BoundDeconstruct
             }
 
             BoundExpression loweredRight = VisitExpression(node.Right);
+            ImmutableArray<BoundExpression> rhsValues;
 
-            // prepare out parameters for Deconstruct
-            var deconstructParameters = node.DeconstructMember.Parameters;
-            var outParametersBuilder = ArrayBuilder<BoundExpression>.GetInstance(deconstructParameters.Length);
-            Debug.Assert(deconstructParameters.Length == node.LeftVariables.Length);
-
-            for (int i = 0; i < numVariables; i++)
+            // get or make right-hand-side values
+            if (node.Right.Type.IsTupleType)
             {
-                var localSymbol = new SynthesizedLocal(_factory.CurrentMethod, deconstructParameters[i].Type, SynthesizedLocalKind.LoweringTemp);
-
-                var localBound = new BoundLocal(syntax,
-                                                localSymbol,
-                                                null,
-                                                deconstructParameters[i].Type
-                                                ) { WasCompilerGenerated = true };
-
-                temps.Add(localSymbol);
-                outParametersBuilder.Add(localBound);
+                rhsValues = AccessTupleFields(node, loweredRight, temps, stores);
+            }
+            else
+            {
+                rhsValues = CallDeconstruct(node, loweredRight, temps, stores);
             }
 
-            var outParameters = outParametersBuilder.ToImmutableAndFree();
-
-            // invoke Deconstruct
-            var invokeDeconstruct = MakeCall(syntax, loweredRight, node.DeconstructMember, outParameters, node.DeconstructMember.ReturnType);
-            stores.Add(invokeDeconstruct);
-
-            // assign from out temps to lhs receivers
-            for (int i = 0; i < numVariables; i++)
+            // assign from rhs values to lhs receivers
+            int numAssignments = node.Assignments.Length;
+            for (int i = 0; i < numAssignments; i++)
             {
                 // lower the assignment and replace the placeholders for source and target in the process
                 var assignmentInfo = node.Assignments[i];
 
                 AddPlaceholderReplacement(assignmentInfo.LValuePlaceholder, lhsReceivers[i]);
-                AddPlaceholderReplacement(assignmentInfo.RValuePlaceholder, outParameters[i]);
+                AddPlaceholderReplacement(assignmentInfo.RValuePlaceholder, rhsValues[i]);
```

Comment:
It looks like this loop interleaves LHS accesses and RHS reads. That does not seem right.
It should be something like:
1) go through all LHS expressions, lower them, and make temps if they are sideeffecting in their lowered form. 
2) go through all RHS while lowering and making assignments to the lowered LHS directly (if does not need temp) or to a corresponding temp made at step1